### ISSUE TITLE
Login fails with interaction_in_progress error

### ISSUE THREAD

**Issue by izikl** [AzureAD/microsoft-authentication-library-for-js#4295]
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Core Library Version

2.18.0

### Wrapper Library

Not Applicable

### Wrapper Library Version

None

### Description

Eventhough, handleRedirectPromise is the first thing that happens in our SPA app, interaction_in_progress is thrown when logging-in. 

`storeAuthStateInCookie` is set to true.

This is the cookie I see in Chrome's dev tools:
![image](https://user-images.githubusercontent.com/8294298/144327003-1f1d4fe9-8cb8-4b49-ad56-ca8491d22040.png)

### Error Message

(with verbose logs enabled) 
![image](https://user-images.githubusercontent.com/8294298/144326787-b4457c5d-0e5d-4460-9399-1ae3440971cc.png)

### Msal Logs

MsalPublicClientApplicationConfig.ts:37 msal: 3: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Verbose - handleRedirectPromise called
MsalPublicClientApplicationConfig.ts:37 msal: 3: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Verbose - getAllAccounts called
MsalPublicClientApplicationConfig.ts:37 msal: 4: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Trace - BrowserCacheManager.getIdTokenCredential: cache hit
MsalPublicClientApplicationConfig.ts:37 msal: 4: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Trace - BrowserCacheManager.getAccessTokenCredential: cache hit
MsalPublicClientApplicationConfig.ts:37 msal: 4: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Trace - BrowserCacheManager.getRefreshTokenCredential: cache hit
MsalPublicClientApplicationConfig.ts:37 msal: 4: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Trace - BrowserCacheManager.getAccessTokenCredential: cache hit
MsalPublicClientApplicationConfig.ts:37 msal: 2: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Info - Emitting event: msal:handleRedirectStart
MsalPublicClientApplicationConfig.ts:37 msal: 3: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Verbose - handleRedirectPromise has been called for the first time, storing the promise
MsalPublicClientApplicationConfig.ts:37 msal: 4: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Trace - BrowserCacheManager.getTemporaryCache: No cache item found in local storage
MsalPublicClientApplicationConfig.ts:37 msal: 3: [Thu, 25 Nov 2021 03:06:42 GMT] : [2ef17b53-b723-4562-932d-a253ca219c99] : msal.js.browser@2.18.0 : Verbose - initializeServerTelemetryManager called
MsalPublicClientApplicationConfig.ts:37 msal: 4: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Trace - BrowserCacheManager.getTemporaryCache: storeAuthStateInCookies set to true, retrieving from cookies
MsalPublicClientApplicationConfig.ts:37 msal: 3: [Thu, 25 Nov 2021 03:06:42 GMT] : [2ef17b53-b723-4562-932d-a253ca219c99] : msal.js.browser@2.18.0 : Verbose - getRedirectResponseHash called
MsalPublicClientApplicationConfig.ts:37 msal: 4: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Trace - BrowserCacheManager.getTemporaryCache: No cache item found in local storage
MsalPublicClientApplicationConfig.ts:37 msal: 4: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Trace - BrowserCacheManager.removeItem: storeAuthStateInCookie is true, clearing item cookie
MsalPublicClientApplicationConfig.ts:37 msal: 3: [Thu, 25 Nov 2021 03:06:42 GMT] : [2ef17b53-b723-4562-932d-a253ca219c99] : msal.js.browser@2.18.0 : Verbose - Hash does not contain known properties, returning cached hash
MsalPublicClientApplicationConfig.ts:37 msal: 2: [Thu, 25 Nov 2021 03:06:42 GMT] : [2ef17b53-b723-4562-932d-a253ca219c99] : msal.js.browser@2.18.0 : Info - handleRedirectPromise did not detect a response hash as a result of a redirect. Cleaning temporary cache.
MsalPublicClientApplicationConfig.ts:37 msal: 4: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Trace - BrowserCacheManager.cleanRequestByInteractionType called
MsalPublicClientApplicationConfig.ts:37 msal: 2: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Info - Emitting event: msal:handleRedirectEnd
MsalPublicClientApplicationConfig.ts:37 msal: 3: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Verbose - loginRedirect called
MsalPublicClientApplicationConfig.ts:37 msal: 3: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Verbose - preflightBrowserEnvironmentCheck started
MsalPublicClientApplicationConfig.ts:37 msal: 3: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Verbose - acquireTokenRedirect called
MsalPublicClientApplicationConfig.ts:37 msal: 3: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Verbose - getAllAccounts called
MsalPublicClientApplicationConfig.ts:37 msal: 4: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Trace - BrowserCacheManager.getIdTokenCredential: cache hit
MsalPublicClientApplicationConfig.ts:37 msal: 4: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Trace - BrowserCacheManager.getAccessTokenCredential: cache hit
MsalPublicClientApplicationConfig.ts:37 msal: 4: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Trace - BrowserCacheManager.getRefreshTokenCredential: cache hit
MsalPublicClientApplicationConfig.ts:37 msal: 4: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Trace - BrowserCacheManager.getAccessTokenCredential: cache hit
MsalPublicClientApplicationConfig.ts:37 msal: 2: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Info - Emitting event: msal:acquireTokenStart
MsalPublicClientApplicationConfig.ts:37 msal: 3: [Thu, 25 Nov 2021 03:06:42 GMT] : [e64d39f1-8604-477f-a90d-b0ef2d7381f9] : msal.js.browser@2.18.0 : Verbose - preflightInteractiveRequest called, validating app environment
MsalPublicClientApplicationConfig.ts:37 msal: 4: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Trace - BrowserCacheManager.getTemporaryCache: storeAuthStateInCookies set to true, retrieving from cookies
MsalPublicClientApplicationConfig.ts:37 msal: 2: [Thu, 25 Nov 2021 03:06:42 GMT] : @azure/msal-browser@2.18.0 : Info - Emitting event: msal:acquireTokenFailure
AuthError.ts:45 
        
       Uncaught (in promise) BrowserAuthError: interaction_in_progress: Interaction is currently in progress. Please ensure that this interaction has been completed before calling an interactive API.  For more visit: aka.ms/msaljs/browser-errors.
    at BrowserAuthError.AuthError [as constructor] (AuthError.ts:45)
    at new BrowserAuthError (BrowserAuthError.ts:159)
    at Function.BrowserAuthError.createInteractionInProgressError (BrowserAuthError.ts:240)
    at RedirectClient.StandardInteractionClient.preflightInteractiveRequest (StandardInteractionClient.ts:206)
    at RedirectClient.<anonymous> (RedirectClient.ts:24)
    at step (viewportRowModelModule.js:4)
    at Object.next (viewportRowModelModule.js:4)
    at viewportRowModelModule.js:4
    at new Promise (<anonymous>)
    at __awaiter (viewportRowModelModule.js:4)
    at RedirectClient.acquireToken (RedirectClient.ts:23)
    at PublicClientApplication.<anonymous> (ClientApplication.ts:205)
    at step (viewportRowModelModule.js:4)
    at Object.next (viewportRowModelModule.js:4)
    at viewportRowModelModule.js:4
    at new Promise (<anonymous>)
    at __awaiter (viewportRowModelModule.js:4)
    at PublicClientApplication.ClientApplication.acquireTokenRedirect (ClientApplication.ts:191)
    at PublicClientApplication.<anonymous> (PublicClientApplication.ts:65)
    at step (viewportRowModelModule.js:4)
    at Object.next (viewportRowModelModule.js:4)
    at viewportRowModelModule.js:4
    at new Promise (<anonymous>)
    at __awaiter (viewportRowModelModule.js:4)
    at PublicClientApplication.loginRedirect (PublicClientApplication.ts:63)
    at MsalAuthenticationProvider._callee5$ (MsalAuthenticationProvider.ts:108)
    at tryCatch (runtime.js:63)
    at Generator.invoke [as _invoke] (runtime.js:293)
    at Generator.next (runtime.js:118)
    at asyncGeneratorStep (asyncToGenerator.js:3)
    at _next (asyncToGenerator.js:25)
    at asyncToGenerator.js:32
    at new Promise (<anonymous>)
    at MsalAuthenticationProvider.<anonymous> (asyncToGenerator.js:21)
    at MsalAuthenticationProvider.login (MsalAuthenticationProvider.ts:24)
    at _callee3$ (index.tsx:63)
    at tryCatch (runtime.js:63)
    at Generator.invoke [as _invoke] (runtime.js:293)
    at Generator.next (runtime.js:118)
    at asyncGeneratorStep (asyncToGenerator.js:3)
    at _next (asyncToGenerator.js:25)

### MSAL Configuration

```javascript
auth: {
        clientId: "f9818e52-50bd-463e-8932-a1650bd3fad2",
        authority: "https://login.microsoftonline.com/organizations", // organizations changes to [tenantId] in some scenarios.
    },
    cache: {
        cacheLocation: 'localStorage',
        storeAuthStateInCookie: true,
    },
```

### Relevant Code Snippets

```javascript
handleRedirectPromise is the first thing that happens when the app loads. 
handleRedirectPromise is awaited. 
loginRedirect is only called after handleRedirectPromise finishes.
```

### Reproduction Steps

I am not sure what causes the cookie msal.interaction.status to be added, but if I manually add the cookie `msal.interaction.status` with the clientId as its value, loading the site again (Which calls handleRedirectPromise) doesn't delete it. 

### Expected Behavior

msal.interaction.status should always be deleted when calling `handleRedirectPromise` (At least for the same clientId). 

### Identity Provider

Azure AD / MSA

### Browsers Affected (Select all that apply)

Chrome, Edge

### Regression

@azure/msal-browser@2.15.0

### Source

Internal (Microsoft)

**Comment by sameerag** [comment]
@izikl Can you please share the code fragment where you are calling `acquireTokenRedirect` and `handleRedirectPromise`?

**Comment by sameerag** [comment]
@izikl The "hash" is being rejected by MSAL in the below error:
>>MsalPublicClientApplicationConfig.ts:37 msal: 3: [Thu, 25 Nov 2021 03:06:42 GMT] : [2ef17b53-b723-4562-932d-a253ca219c99] : msal.js.browser@2.18.0 : Verbose - Hash does not contain known properties, returning cached hash

We clear all the temporary cache after the failure, and the cookies should have been cleared too.
Can you try setting `storeAuthStateInCookie: false` and let us know if you are still seeing the issue?

**Comment by izikl** [comment]
@sameerag 
> Can you please share the code fragment where you are calling acquireTokenRedirect and handleRedirectPromise?
1. [call to handleRedirectPromise](https://dev.azure.com/msazure/One/_git/Azure-Kusto-WebUX?path=/kustoweb/src/AuthenticationProviders/MsalAuthProvider/MsalAuthenticationProvider.ts&version=GBmaster&line=58&lineEnd=59&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents)  
2. [Trigger of handleRedirectPromise](https://dev.azure.com/msazure/One/_git/Azure-Kusto-WebUX?path=/kustoweb/src/utils/Login/loginManager.tsx&version=GBmaster&line=92&lineEnd=93&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents) - This is the first line (related to authentication) in the app. 
3. loginRedirect is called from [both loginWithTenantFromQuery and basicLoginFlow](https://dev.azure.com/msazure/One/_git/Azure-Kusto-WebUX?path=/kustoweb/src/utils/Login/loginManager.tsx&version=GBmaster&line=95&lineEnd=103&lineStartColumn=9&lineEndColumn=10&lineStyle=plain&_a=contents) 
4. acquireTokenRedirect is called [here](https://dev.azure.com/msazure/One/_git/Azure-Kusto-WebUX?path=/kustoweb/src/AuthenticationProviders/MsalAuthProvider/MsalAuthenticationProvider.ts&version=GBmaster&line=45&lineEnd=46&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents) after acquireTokenSilently fails with InteractionRequiredAuthError.

In short:
1. handleRedirectPromise - is the first thing that is executed when the app loads. (it is awaited)
2. loginRedirect - might be called next if there is not cached token
3. acquireTokenRedirect - called after a user is already logged in, and whenever a token can't be fetched silently.

**Comment by izikl** [comment]
> We clear all the temporary cache after the failure, and the cookies should have been cleared too.

Only cookies with prefix `msal.<client ID>` are deleted:
[const cookiePrefix = `${Constants.CACHE_PREFIX}.${this.clientId}`;](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/8afb0be101c2ff51533fce867791efdba6d88ee6/lib/msal-browser/src/cache/BrowserCacheManager.ts#L631)

However, during my tests the cookie was set with the `msal.` prefix and not with `msal.<client ID>`:
![image](https://user-images.githubusercontent.com/8294298/144582164-7dd7cbc2-db08-499d-bc48-85e07f5e7cc4.png)

Here is the line that sets the interaction.status key. Note that addInstanceId  is set to **false**: 
[this.setTemporaryItem(this.generateCacheKey(TemporaryCacheKeys.INTERACTION_STATUS, **false**), this.clientId);](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/48b4e0e033ce922a751d698d254e9455d81992e6/lib/msal-core/src/cache/AuthCache.ts#L336)

Which prevents the client ID from being added:
[return addInstanceId ? `${Constants.cachePrefix}.${this.clientId}.${key}` : `${Constants.cachePrefix}.${key}`;](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/48b4e0e033ce922a751d698d254e9455d81992e6/lib/msal-core/src/cache/AuthCache.ts#L94)

I believe that this is why the cookie was never deleted. 

> Can you try setting storeAuthStateInCookie: false and let us know if you are still seeing the issue?

I couldn't repro it with `storeAuthStateInCookie: false`. I never saw an interaction.status key in the local storage and adding it manually never threw interaction_in_progress error, so I am not sure how much value those tests has.

**Comment by samushenkov** [comment]
Hello there. We are getting totally the same exception - interaction_in_progress.

config:
```
{
  storeAuthStateInCookie: true,
  cacheLocation: 'memoryStorage'
}
```

There is a chance to break an application until end user "clears site data". It happens if user closes page during redirect process (from b2c to our app with auth state provided). MSAL adds **msal.interaction.status** cookie and doesn't delete it (user closes page before).

First of all we start working with handleRedirectPromise method that doesn't remove mentioned cookie. And then we use acquireTokenRedirect that throws exception above.

**Comment by tomlagier** [comment]
Not sure if this is related, but I just started seeing this error with the following config. It had seemingly worked OK for the last couple of weeks, not sure if we're seeing an edge case, some service instability, or just plain misconfiguration that didn't show up until now:

`authConfig.ts`
```
export const msalConfig: Configuration = {
  auth: { ... },
  cache: {
    cacheLocation: "localStorage", // Configures cache location. "sessionStorage" is more secure, but "localStorage" gives you SSO between tabs.
    storeAuthStateInCookie: false, // If you wish to store cache items in cookies as well as browser cache, set this to "true".
  },
};
```

`auth.tsx.`
```
// Set up MSAL application instance. We do this at parse time so we can share this object
// outside of hooks. Useful for `getAuthenticatedHeaders` helper function.
const pca = new PublicClientApplication(msalConfig);

function B2CAuthProvider({ children }: B2CAuthProviderProps) {
  // Use custom routing for a smoother post-redirect experience
  const router = useRouter();
  const navigationClient = new CustomNavigationClient(router);
  pca.setNavigationClient(navigationClient);

  // Set up initial render hooks
  useEffect(() => {
    // Default to using the first account if no account is active on page load
    if (!pca.getActiveAccount() && pca.getAllAccounts().length > 0) {
      // Account selection logic is app dependent. Adjust as needed for different use cases.
      pca.setActiveAccount(pca.getAllAccounts()[0]);
    }

    // Optional - This will update account state if a user signs in from another tab or window
    pca.enableAccountStorageEvents();

    // Set the active account whenever the user logs in to a new account or fetches a new token
    pca.addEventCallback(event => {
      if (
        event.eventType === EventType.LOGIN_SUCCESS &&
        event.payload &&
        "homeAccountId" in event.payload
      ) {
        pca.setActiveAccount(event.payload);
      }

      if (
        [
          EventType.SSO_SILENT_SUCCESS,
          EventType.ACQUIRE_TOKEN_SUCCESS,
        ].includes(event.eventType) &&
        event.payload &&
        "account" in event.payload &&
        event.payload.account
      ) {
        pca.setActiveAccount(event.payload.account!);
      }
    });
  }, []);

  return <MsalProvider instance={pca}>{children}</MsalProvider>;
}

// Check if a user is logged in. If not, attempt to redirect.
export function useAuthenticationRequired() {
  // Attempt to silently login, if user already has a session on the server
  const { error, login } = useMsalAuthentication(
    InteractionType.Silent,
    loginRequest
  );
  const { inProgress } = useMsal();

  // It failed, use the redirect flow to login
  useEffect(() => {
    if (error) {
      login(InteractionType.Redirect, loginRequest);
    }
  }, [error, login]);
}
```

`some-page.tsx`
```
export function Page() {
  useAuthenticationRequired();
  return <div />;
}
```

`@azure/msal-react` v1.1.1, `@azure/msal-browser` v2.19.0.

**Comment by izikl** [comment]
@samushenkov until this bug is fixed, I workaround this bug by deleting the cookie `msal.interaction.status` immediately after the call to `handleRedirectPromise`. 

```typescript
document.cookie = 'msal.interaction.status=;path=/;expires=Thu, 01 Jan 1970 00:00:01 GMT'
```

**Comment by samushenkov** [comment]
Yep, did it too.

**Comment by jasonnutter** [comment]
This should be address by #4314

**Comment by izikl** [comment]
@jasonnutter I don't think this it is going to fix it. 

The cookie deletion always assumes this prefix `msal.<client ID>`, as shown here:
https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/8afb0be101c2ff51533fce867791efdba6d88ee6/lib/msal-browser/src/cache/BrowserCacheManager.ts#L631

However, the cookie is saved without client ID.  

(I am describing it in more details here https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/4295#issuecomment-985379347)

**Comment by jasonnutter** [comment]
@izikl Right, which is why the PR adds a new invocation of `this.setInteractionInProgress(false);`, which properly handles the cache key.

**Comment by izikl** [comment]
@jasonnutter oh yeah. My mistake.
